msgraph/serviceprincipals.go

```diff
@@ -348,6 +348,136 @@ func (c *ServicePrincipalsClient) RemoveOwners(ctx context.Context, servicePrinc
 	return status, nil
 }
 
+// AssignClaimsMappingPolicy assigns a claimsMappingPolicy to a servicePrincipal
+func (c *ServicePrincipalsClient) AssignClaimsMappingPolicy(ctx context.Context, servicePrincipal *ServicePrincipal) (int, error) {
+	var status int
+
+	if servicePrincipal.ID == nil {
+		return status, errors.New("cannot update service principal with nil ID")
+	}
+	if servicePrincipal.ClaimsMappingPolicies == nil {
+		return status, errors.New("cannot update service principal with nil ClaimsMappingPolicies")
+	}
+
+	for _, policy := range *servicePrincipal.ClaimsMappingPolicies {
+		// don't fail if an owner already exists
+		checkPolicyAlreadyExists := func(resp *http.Response, o *odata.OData) bool {
+			if resp != nil && resp.StatusCode == http.StatusBadRequest && o != nil && o.Error != nil {
+				return o.Error.Match(odata.ErrorAddedObjectReferencesAlreadyExist)
+			}
+			return false
+		}
+
+		path := odata.Id(fmt.Sprintf("https://graph.microsoft.com/v1.0/policies/claimsMappingPolicies/%s", *policy.ID))
+		body, err := json.Marshal(ClaimsMappingPolicy{ODataId: &path})
```

Comment:
Whilst the handling of $ref endpoints is a bit clunky at the moment, the construction of the OData URI can be left to the odata package and this can be reduced to:

```suggestion
		body, err := json.Marshal(DirectoryObject{ODataId: policy.ODataId})
```

This does require that the `ODataId` field is set on the model, however this can be accomplished by embedding the `DirectoryObject` struct and requesting the full metadata in the `ClaimsMappingPolicyClient` (see next comments).